modin/core/dataframe/algebra/groupby.py

```diff
@@ -290,6 +291,21 @@ def caller(
             )
         assert axis == 0, "Can only groupby reduce with axis=0"
 
+        # The bug only occurs in the case of Categorical 'by', so we might want to check whether any of
+        # the 'by' dtypes is Categorical before going into this branch, however triggering 'dtypes'
+        # computation if they're not computed may take time, so we don't do it
+        if not groupby_args.get("sort", True) and isinstance(by, type(query_compiler)):
+            ErrorMessage.missmatch_with_pandas(
+                operation="df.groupby(categorical_by, sort=False)",
+                message=(
+                    "the groupby keys will be sorted anyway, although the 'sort=False' was passed. "
+                    "See the following issue for more details: "
+                    "https://github.com/modin-project/modin/issues/3571"
+                ),
+            )
+            groupby_args = groupby_args.copy()
```

Comment:
Why do we copy if we're storing the value in the same variable?

Reply:
`groupby_args` is an user's object, we don't want to change it:
```python
>>> user_dict = {"sort": False}
>>> def our_groupby(groupby_params):
...     groupby_params["sort"] = True
...     # Do the stuff...
...
>>> user_dict
{'sort': False}
>>> our_groupby(user_dict)
>>> user_dict
{'sort': True}
```